module-4/nmap/nmap.py

```diff
@@ -0,0 +1,80 @@
+import socket
+import argparse
+import concurrent.futures
+
+# Common services for the first 100 ports
+COMMON_PORTS = {
+    20: "FTP data",
+    21: "FTP control",
+    22: "SSH",
+    23: "Telnet",
+    25: "SMTP",
+    53: "DNS",
+    80: "HTTP",
+    110: "POP3",
+    143: "IMAP",
+    443: "HTTPS",
+    993: "IMAPS",
+    995: "POP3S",
+    # ... continue up to port 100 as needed
+}
+
+def resolve_host(host):
+    try:
+        ip = socket.gethostbyname(host)
+        print(f"[INFO] IP address for {host} is {ip}")
+        return ip
+    except socket.gaierror:
+        print(f"[ERROR] Could not resolve host {host}.")
+        return None
+
+def scan_port(ip, port):
+    with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
+        s.settimeout(1)
+        result = s.connect_ex((ip, port))
+        if result == 0:
+            service_name = COMMON_PORTS.get(port, "Unknown Service") if port <= 100 else "Unknown"
+            print(f"Port {port} is open (Service: {service_name})")
+            return port, True, service_name
+        else:
+            print(f"Port {port} is closed")
+            return port, False, None
+
+def port_scan(ip, port_range):
+    open_ports = []
+    with concurrent.futures.ThreadPoolExecutor(max_workers=100) as executor:
+        future_to_port = {executor.submit(scan_port, ip, port): port for port in port_range}
+        for future in concurrent.futures.as_completed(future_to_port):
+            port, is_open, service_name = future.result()
+            if is_open:
+                open_ports.append((port, service_name))
+    return open_ports
+
+def parse_port_range(port_range_str):
+    if port_range_str == "-":
+        return range(1, 65536)
+    else:
+        min_port, max_port = map(int, port_range_str.split("-"))
+        return range(min_port, max_port + 1)
```

Comment:
I would do the function logic the other way around:

```python
min_port, max_port = port_range_str.split("-")
try:
    if min_port == "" or max_port == "":
        raise ValueError("Invalid port values provided")
    else:
        min_port = int(min_port)
        max_port = int(max_port)
    if min_port < 1 or max_port > 65536:
        raise ValueError("Illegal value for port boundary provided")
except ValuesError as ve:
    min_port = 0
    max_port = 65536
finally:
    return range(min_port, max_port + 1)
```

**Reasoning:**

- Only one return-statement ensures readability and avoids confusing with pathing
- ensure provided values are in correct range/format
- includes error handling